src/txrequest.cpp

```diff
@@ -0,0 +1,724 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txrequest.h>
+
+#include <crypto/siphash.h>
+#include <net.h>
+#include <primitives/transaction.h>
+#include <random.h>
+#include <uint256.h>
+#include <util/memory.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+
+#include <assert.h>
+
+namespace {
+
+/** The various states a (txhash,peer) pair can be in.
+ *
+ * Note that CANDIDATE is split up into 3 substates (DELAYED, BEST, READY), allowing more efficient implementation.
+ * Also note that the sorting order of ByTxHashView relies on the specific order of values in this enum.
+ *
+ * Expected behaviour is:
+ *   - When first announced by a peer, the state is CANDIDATE_DELAYED until reqtime is reached.
+ *   - Announcements that have reached their reqtime but not been requested will be either CANDIDATE_READY or
+ *     CANDIDATE_BEST
```

Comment:
I didn't figure this during previous reviews, but this doesn't underscore that CANDIDATE_READY expiration time is never if it's never promoted to REQUESTED or the expected transaction is never received. Just to prevent a future change introducing a expiration time for then and thus a strong dependency order where a first malicious request could interfere with honest CANDIDATE_READY.

"CANDIDATE_READY announcements may stay in this state until they're promoted to REQUESTED or the announced transaction is no longer needed"

Reply:
I can't understand your first paragraph.

Your suggested text would be incorrect, as they're expected to first transition to CANDIDATE_BEST, before becoming REQUESTED. Feel free to suggest something that includes that.

Reply:
Just wanted to add a comment saying that _READY expiration time is never, if it doesn't get promoted to _BEST or the announced transaction is no longer needed.

Right forget a step,  "CANDIDATE_READY announcements don't expire until they're promoted to _BEST or the announced transaction is no longer needed" better?

Reply:
Ok, added some more comments.